File: utils/variables.py
from typing import Union, Callable
import json
from .timers import PausableTimer


class Variables:
    """A variables dictionary with some extra functionalities, like state backup and supervise intervaled callback.
    
    Args:
        variables: a dictionary with variables.
        enabled: flag to enable or disable variables control.
        interval: max wait time in seconds for a response.
        supervise: a callback to be executed after wait time passes.

    Example:
        variables = Variable({
            'var1': 1, # type: int
            'var2': 3.14 # type: float
            'var3': 'active' # type: str
            'var4': False, # type: bool
        })
    """
    def __init__(
        self, 
        variables: dict = {}, 
        enabled: bool = True,
        interval: Union[float, int] = 2,
        supervise: Callable = None,
    ):
        self.variables = variables
        self.backup = variables.copy()
        self.enabled = enabled
        self.streamingStatus = False
        self.timer = PausableTimer(interval, supervise)

    def __len__(self):
        return len(self.variables)

    def __str__(self):
        return str(self.variables)
    
    def __getitem__(self, key):
        return self.variables[key]
    
    def __setitem__(self, key: str, value):
        self.variables[key] = value
    
    def isEnabled(self):
        """Checks if variables workflow is enabled."""
        return self.enabled

    def setEnabled(self, value: bool = True):
        """Updates the enable status"""
        self.enabled = value

    def restore(self):
        """Restores the variables backup."""
        self.variables = self.backup.copy()

    def set(self, key: str, value, backup: bool = True, streamingStatus: bool = False):
        """Updates a variable value"""
        if backup:
            self.backup = self.variables.copy()
        self.variables[key] = value
        self.setStreamingStatus(streamingStatus)
    
    def get(self, key: str):
        """Returns a specific variable value."""
        return self.variables[key]

    def values(self):
        """Get the variables values on dict format."""
        return self.variables

    def json(self):
        """Returns the variables dict as JSON string."""
        return json.dumps(self.variables)

    def update(self, data: Union[str, dict] = {}):
        """Updates the variables values."""
        if isinstance(data, str):
            data = json.loads(data)
        self.variables = data
        self.backup = dict(self.variables)

    def streamed(self):
        """Returns the current streaming status."""
        return self.streamingStatus

    def setStreamingStatus(self, value: bool = True):
        """Updates the streaming status."""
        self.streamingStatus = value

    def streamedSucessfully(self):
        """Should be called when variables streaming were successfully."""
        self.setStreamingStatus(True)
        self.timer.resume(now=True)

    def waitResponse(self):
        """Starts the supervise loop (timer)."""
        self.timer.resume(now=False)

    def checkStreamingFail(self):
        """If variables were not streamed restores the variables backup."""
        if not self.streamed():
            self.restore()

    def resetStreamingStatus(self):
        """Restores streaming status and pauses the check timer."""
        self.setStreamingStatus(False)
        self.timer.pause(reset=True)
    
    def stop(self):
        """Stops the check timer."""
        self.timer.stop()

File: public/js/utils/variables.js
import { PausableTimer } from './timers.js';


export class Variables {
    /**
     * Creates an inteligent variables manager (JSON).
     * @param {object} variables - a json with variables names and values.
     */
    constructor(variables={}, enabled=true, interval=2.0, supervise){
        this.variables = variables;
        this.backup = {...variables};
        this.enabled = enabled;
        this.streamingStatus = false;
        this.timer = new PausableTimer(interval, supervise);
    }

    toString(){
        return this.variables;
    }

    /**
     * Restores the variables values using the backup.
     */
    restore(){
        this.variables = {...this.backup};
    }

    /** 
     * Returns the variables values.
    */
    values(){
        return this.variables;
    }

    /**
     * Updates a specific variable value
     * @param {string} key - the variable name
     * @param {*} value - a value for the current variable
     * @param {boolean} backup - save the previous variables values?
     * @param {boolean} streamingStatus - flag
     */
    set = (key, value, backup = true, streamingStatus = false) => {
        if(backup){
            this.backup = {...this.variables};
        }
        this.variables[key] = value;
        this.setStreamingStatus(streamingStatus);
    }
    
    /**
     * Returns a value of the variables JSON given a key
     * @param {string} key - the current variable name
     */
    get = (key) => {
        return this.variables[key];
    }

    /** Checks if variables is enabled */
    isEnabled = () => {
        return this.enabled;
    }

    /**
     * Updates the updated status.
     * @param {boolean} value.
     */
    setStreamingStatus = (value) => {
        this.streamingStatus = value;
    }
    
    /** Stringify variables */
    json = () => {
        return JSON.stringify(this.variables)
    }

    /** 
     * Updates variables values.
     * @param {string | object} data - a json string with variables.
    */
    update = (data) => {
        let variables = data;
        if (typeof variables === 'string'){
            variables = JSON.parse(variables);
        }
        this.variables = {...variables};
        this.backup = {...variables};
        this.setStreamingStatus(true);
    }

    /** Returns the current streaming status */
    streamed = () => {
        return this.streamingStatus;
    }

    /**
     * Updates the streaming status.
     * @param {boolean} value - the new streaming status value.
     */
    setStreamingStatus = (value=true) => {
        this.streamingStatus = value;
        this.timer.reset(true);
    }

    /**
     * Waits for a response
     */
    waitResponse = () => {
        this.timer.resume(false);
    }

    /** Checks if variables were streamed, else restores their last values. */
    checkStreamingFail = () => {
        if(!this.streamed()){
            this.restore();
        }
    }

    /** Resets the streaming status. */
    resetStreamingStatus = () => {
        this.setStreamingStatus(false);
        this.timer.pause()
    }

    streamedSucessfully = () => {
        console.log("streamed succesfully!");
        this.setStreamingStatus(true);
        this.timer.resume(true);
    }

    /** Stops secondaries processes like timeouts or intervals. */
    stop = () => {
        this.timer?.stop()
    }


}
